pkg/healthz/healthz_test.go

```diff
@@ -0,0 +1,359 @@
+					"retryPeriod", retryPeriod,
+				)
+				g.Go(func() error {
+					defer log.Info("Leader election go-routine finished")
+					leaderelection.RunOrDie(gCTX, leaderelection.LeaderElectionConfig{
+						LeaseDuration: leaseDuration,
+						RenewDeadline: renewDeadline,
+						RetryPeriod:   retryPeriod,
+						Callbacks: leaderelection.LeaderCallbacks{
+							OnStartedLeading: func(context.Context) {
+								log.Info("leaderelection.LeaderCallbacks.OnStartedLeading")
+							},
+							OnStoppedLeading: func() {
+								log.Info("leaderelection.LeaderCallbacks.OnStoppedLeading")
+							},
+							OnNewLeader: func(identity string) {
+								log.Info("leaderelection.LeaderCallbacks.OnNewLeader", "identity", identity)
+								if tc.in.onNewLeaderHook != nil {
+									tc.in.onNewLeaderHook(&tc.in)
+								}
+								close(leaderElected)
+							},
+						},
+						Lock:     tc.in.resourceLock,
+						WatchDog: s.LeaderHealthzAdaptor,
+					})
+					return nil
+				})
+			}
+
+			log.Info("Starting healthz server go-routine")
+			g.Go(func() error {
+				defer log.Info("Healthz server go-routine finished")
+				return s.Start(gCTX, l)
+			})
+
+			if tc.in.leaderElectionEnabled {
+				log.Info("Waiting for a LeaderElector to know the current leader before polling liveness endpoint")
+				<-leaderElected
+			}
+
+			const (
+				pollingInterval = 500 * time.Millisecond
+				pollingTimeout  = 3 * time.Second
+			)
+			log.Info(
+				"Polling liveness endpoint",
+				"url", livezURL,
+				"interval", pollingInterval,
+				"timeout", pollingTimeout,
+			)
+			var (
+				lastResponseCode int
+				lastResponseBody string
+			)
+			assert.Eventually(t, func() bool {
+				resp, err := http.Get(livezURL)
+				require.NoError(t, err)
+				defer func() {
+					require.NoError(t, resp.Body.Close())
+				}()
+				bodyBytes, err := ioutil.ReadAll(resp.Body)
+				require.NoError(t, err)
+
+				lastResponseCode = resp.StatusCode
+				lastResponseBody = string(bodyBytes)
+
+				log.Info("liveness-probe", "response-code", lastResponseCode, "response-body", lastResponseBody)
+
+				return tc.out.responseCode == lastResponseCode && tc.out.responseBody == lastResponseBody
+			}, pollingTimeout, pollingInterval)
+
+			assert.Equal(t, tc.out.responseBody, lastResponseBody)
+			assert.Equal(t, tc.out.responseCode, lastResponseCode)
+			cancel()
+			require.NoError(t, g.Wait())
+		})
+	}
+}
+
+// fakeResourceLock implements resourcelock.Interface sufficiently to simulate:
+// * successful acquisition of the leader election lock by the local node,
+// * current possession of the leader election lock by a remote node, and
+// * failures in leader election which cause the `LeaderElection.Run` function to exit.
+//
+// The intention is to be able to test the behavior of the
+// LeaderElectionHealthzAdaptor under those circumstances.
+type fakeResourceLock struct {
+	record      *resourcelock.LeaderElectionRecord
+	getError    error
+	updateError error
+}
```

Comment:
[controller-runtime also has a fake resourcelock](https://github.com/kubernetes-sigs/controller-runtime/blob/0f5aba94e2f617298270f561ed8de2b377be204a/pkg/leaderelection/fake/leader_election.go) but it didn't allow me to inject errors.